lib/src/Pos/DB/Block.hs

```diff
@@ -72,180 +47,92 @@ import           System.IO.Error (IOError, isDoesNotExistError)
+import           Pos.DB.GState.Common (getTipSomething)
 import           Pos.DB.Pure (DBPureVar, MonadPureDB, atomicModifyIORefPure, pureBlockIndexDB,
                               pureBlocksStorage)
 import           Pos.DB.Rocks (MonadRealDB, blockDataDir, getBlockIndexDB, getNodeDBs, rocksDelete,
                                rocksPutBi)
 import           Pos.DB.Sum (MonadDBSum, eitherDB)
 import           Pos.Delegation.Types (DlgUndo (..))
-import           Pos.Ssc (SscBlock)
-import           Pos.Ssc.Util (toSscBlock)
-import           Pos.Util (Some (..), maybeThrow)
-import           Pos.Util.Chrono (NewestFirst (..))
+import           Pos.Util.Util (eitherToThrow)
 
 ----------------------------------------------------------------------------
--- Implementations for 'MonadRealDB'
+-- BlockDB related methods
 ----------------------------------------------------------------------------
 
--- Get block with given hash from Block DB.  This function has too
--- strict constraint, consider using 'blkGetBlock'.
+getUndo :: MonadDBRead m => HeaderHash -> m (Maybe Undo)
+getUndo x = dbGetRawUndo x >>= \case
+    Nothing      -> pure Nothing
+    Just rawUndo -> eitherToThrow $ first DBMalformed $ decodeFull $ unRawUndo rawUndo
 
-getBlock
-    :: forall ctx m. (HasConfiguration, MonadRealDB ctx m)
-    => HeaderHash -> m (Maybe Block)
-getBlock = blockDataPath >=> getData
+-- | Convenient wrapper which combines 'dbGetBlock' and 'dbGetUndo' to
+-- read 'Blund'.
+getBlund :: MonadDBRead m => HeaderHash -> m (Maybe (Block, Undo))
+getBlund x =
+    runMaybeT $
+    (,) <$> MaybeT (getBlock x)
+        <*> MaybeT (getUndo x)
+
+putBlund :: MonadDB m => Blund -> m ()
+putBlund = dbPutRawBlund . fmap (RawUndo . serialize')
+
+-- | Get 'Block' corresponding to tip.
+getTipBlock :: MonadDBRead m => m Block
+getTipBlock = getTipSomething "block" getBlock
+
+----------------------------------------------------------------------------
+-- Implementations for 'MonadRealDB'
+----------------------------------------------------------------------------
 
--- | Returns header of block that was requested from Block DB.
-blkGetHeader
-    :: (HasConfiguration, MonadDBRead m)
-    => HeaderHash -> m (Maybe BlockHeader)
-blkGetHeader = dbGetBi BlockIndexDB . blockIndexKey
+-- Get serialization of a block with given hash from Block DB.
+getRawBlock
+    :: forall ctx m. (HasConfiguration, MonadRealDB ctx m)
+    => HeaderHash -> m (Maybe ByteString)
+getRawBlock = blockDataPath >=> getData
 
--- Get undo data for block with given hash from Block DB. This
--- function has too strict constraint, consider using 'blkGetUndo'.
-getUndo :: (HasConfiguration, MonadRealDB ctx m) => HeaderHash -> m (Maybe Undo)
-getUndo = undoDataPath >=> getData
+-- Get serialization of an undo data for block with given hash from Block DB.
+getRawUndo :: (HasConfiguration, MonadRealDB ctx m) => HeaderHash -> m (Maybe ByteString)
+getRawUndo = undoDataPath >=> getData
```

Comment:
Once again, how is it supposed to work? `getData` performs `decodeFull`, but for `getRaw*` you don't need to do any decoding.